### PR TITLE
executor: fix bug when using IndexMerge in transaction

### LINKED ISSUE #30685: UnionScan + IndexMerge,  query with or missing record in transaction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set tidb_enable_index_merge =1;
CREATE TABLE `t2` (
  `c_int` int(11) NOT NULL,
  `c_str` varchar(40) DEFAULT NULL,
  PRIMARY KEY (`c_int`) /*T![clustered_index] CLUSTERED */,
  KEY `c_int_2` (`c_int`),
  KEY `c_str` (`c_str`)
);
insert into t2 values( 7,'infallible feistel');
begin;
update t2  set c_int=5;
select * from t2 where c_int between 4 and 8 or c_str is null;
```

### 2. What did you expect to see? (Required)

```
mysql> select * from t2 where c_int between 4 and 8 or c_str is null;
+-------+--------------------+
| c_int | c_str              |
+-------+--------------------+
|     5 | infallible feistel |
+-------+--------------------+
1 row in set (0.06 sec)
```

### 3. What did you see instead (Required)

```
mysql> select * from t2 where c_int between 4 and 8 or c_str is null;
Empty set (0.00 sec)

mysql> explain select * from t2 where c_int between 4 and 8 or c_str is null;        
+----------------------------------+---------+-----------+------------------------------+-------------------------------------------------------------------------------+
| id                               | estRows | task      | access object                | operator info                                                                 |
+----------------------------------+---------+-----------+------------------------------+-------------------------------------------------------------------------------+
| UnionScan_6                      | 1.00    | root      |                              | or(and(ge(test2.t2.c_int, 4), le(test2.t2.c_int, 8)), isnull(test2.t2.c_str)) |
| └─IndexMerge_16                  | 1.00    | root      |                              |                                                                               |
|   ├─TableRangeScan_13(Build)     | 1.00    | cop[tikv] | table:t2                     | range:[4,8], keep order:false, stats:pseudo                                   |
|   ├─IndexRangeScan_14(Build)     | 1.00    | cop[tikv] | table:t2, index:c_str(c_str) | range:[NULL,NULL], keep order:false, stats:pseudo                             |
|   └─TableRowIDScan_15(Probe)     | 1.00    | cop[tikv] | table:t2                     | keep order:false, stats:pseudo                                                |
+----------------------------------+---------+-----------+------------------------------+-------------------------------------------------------------------------------+
5 rows in set (0.00 sec)

mysql> commit;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t2 where c_int between 4 and 8 or c_str is null;        
+-------+--------------------+
| c_int | c_str              |
+-------+--------------------+
|     5 | infallible feistel |
+-------+--------------------+
1 row in set (0.01 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.4.0-alpha-383-g755553660
Edition: Community
Git Commit Hash: 755553660e6036ead13d1f0aa271eaf59d939fbd
Git Branch: master
UTC Build Time: 2021-12-12 12:39:48
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #30685

Problem Summary: 
1. We enable usage of IndexMerge in transaction in #29875, which add  `memIndexMergeReader` . And it will use kvRanges which are built in IndexMergeReaderExecutor.Open().
2. In memIndexMergeReader, kvRange for partial memTableReader is nil when table's IsPKHandle is true. So we got bug here.

### What is changed and how it works?
Extract kvRange for partial TableReader when open IndexMergeReader.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/index_merge_reader_test.go [hunk 1]
```diff
@@ -273,23 +282,67 @@ func (s *testSuite1) TestIndexMergeInTransaction(c *C) {
...
 	res.Check(testkit.Rows())
-	res = tk.MustQuery("select /*+ use_index_merge(t1) */ * from t1 where (pk < 20 or c2 < 20) and c3 < 20;")
+
+	res = tk.MustQuery("select /*+ use_index_merge(t1) */ * from t1 where (pk < -1 or c2 < 20) and c3 < 20;")
+	res.Check(testkit.Rows())
+	res = tk.MustQuery("select /*+ use_index_merge(t1) */ * from t1 where (pk < 20 or c2 < -1) and c3 < 20;")
 	res.Check(testkit.Rows())
 	tk.MustExec("commit;")
+
+	// This is a case generated by sqlgen to test if clusted index is ok.
```
Comment (cfzjywxk) [inline executor/index_merge_reader_test.go:319]:
```suggestion
	// This is a case generated by sqlgen to test if clustered index is ok.
```

Comment (cfzjywxk) [reply executor/index_merge_reader_test.go:319]: Maybe we could add some compound primary clustered index cases with union scan executor.

Comment (guo-shaoge) [reply executor/index_merge_reader_test.go]: OK. I will add more case to test:
1. Compoud primary clustered key(int col + other col)
2. Single primary key, but is string or other type.

**Review by tiancaiamao — APPROVED** [review]

**Comment by tiancaiamao** [comment]
If we can re-design the coprocessor protocol, and add the raw row key information to the chunk row,
a lot of bugs can be avoid and the code can be simpified a lot.

The UnionScan and SelectLock executor both rely on the raw key of the chunk row... re-calculate the raw key is source of bugs.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Comment by cfzjywxk** [comment]
> If we can re-design the coprocessor protocol, and add the raw row key information to the chunk row, a lot of bugs can be avoid and the code can be simpified a lot.
> 
> The UnionScan and SelectLock executor both rely on the raw key of the chunk row... re-calculate the raw key is source of bugs.

@tiancaiamao The `UnionScan` operator may need to handle not just rowkeys but also unique index keys, this is not handled by now either and leads to several werid behaviours..

**Pull request merged**